test/e2e/e2e_test.go

```diff
@@ -15,106 +15,240 @@
+					return err
 				}
+
 				return nil
+			}).WithTimeout(15 * time.Minute).WithPolling(10 * time.Second).Should(Succeed())
+		})
+	})
+
+	Context("AWS Templates", func() {
+		var (
+			kc          *kubeclient.KubeClient
+			deleteFunc  func() error
+			clusterName string
+			err         error
+		)
+
+		BeforeAll(func() {
+			By("ensuring AWS credentials are set")
+			kc, err = kubeclient.NewFromLocal(namespace)
+			ExpectWithOffset(2, err).NotTo(HaveOccurred())
+			ExpectWithOffset(2, kc.CreateAWSCredentialsKubeSecret(context.Background())).To(Succeed())
+		})
+
+		AfterEach(func() {
+			// If we failed collect logs from each of the affiliated controllers
+			// as well as the output of clusterctl to store as artifacts.
+			if CurrentSpecReport().Failed() {
+				By("collecting failure logs from controllers")
+				collectLogArtifacts(kc, clusterName, deployment.ProviderAWS, deployment.ProviderCAPI)
 			}
-			EventuallyWithOffset(1, verifyControllerUp, time.Minute, time.Second).Should(Succeed())
 
+			// Delete the deployments if they were created.
+			if deleteFunc != nil {
+				By("deleting the deployment")
+				err = deleteFunc()
+				Expect(err).NotTo(HaveOccurred())
+			}
+
+			// Purge the AWS resources, the AfterAll for the controller will
+			// clean up the management cluster.
+			By("nuking remaining AWS resources")
+			err = os.Setenv("CLUSTER_NAME", clusterName)
+			Expect(err).NotTo(HaveOccurred())
+			cmd := exec.Command("make", "dev-aws-nuke")
+			_, err := utils.Run(cmd)
+			ExpectWithOffset(2, err).NotTo(HaveOccurred())
 		})
+
+		for _, template := range []deployment.Template{deployment.TemplateAWSStandaloneCP, deployment.TemplateAWSHostedCP} {
+			It(fmt.Sprintf("should work with an AWS provider and %s template", template), func() {
+				if template == deployment.TemplateAWSHostedCP {
+					// TODO: Create AWS resources for hosted control plane.
+					Skip("AWS hosted control plane not yet implemented")
+				}
+
+				By("creating a Deployment")
+				d := deployment.GetUnstructuredDeployment(deployment.ProviderAWS, template)
+				clusterName = d.GetName()
+
+				deleteFunc, err = kc.CreateDeployment(context.Background(), d)
+				Expect(err).NotTo(HaveOccurred())
+
+				By("waiting for infrastructure providers to deploy successfully")
+				Eventually(func() error {
+					return deployment.VerifyProviderDeployed(context.Background(), kc, clusterName)
+				}).WithTimeout(30 * time.Minute).WithPolling(10 * time.Second).Should(Succeed())
+
+				By("verify the deployment deletes successfully")
+				err = deleteFunc()
+				Expect(err).NotTo(HaveOccurred())
+				Eventually(func() error {
+					return deployment.VerifyProviderDeleted(context.Background(), kc, clusterName)
+				}).WithTimeout(10 * time.Minute).WithPolling(10 * time.Second).Should(Succeed())
+			})
+		}
 	})
 })
+
+func verifyControllerUp(kc *kubeclient.KubeClient, labelSelector string, name string) error {
```

Comment:
I would use higher level resources (like deployments) + some controllers can have more than one pod (Azure).
Also I believe CAPI operator objects could be verified as well (or at least gathered for further analysis).

Reply:
This was a direct copy of the existing test, but I'm happy to fix it.

Reply:
Ah this is from a part which was generated. I guess we should fix it, but probably not in this PR I will not insist on that

Reply:
> Also I believe CAPI operator objects could be verified as well (or at least gathered for further analysis).

I agree, but it's a bit of scope creep for the original issue, I fixed this test because I was trying to get away from the use of `exec.Command` and some changes that came with the rebase when `capi-operator` was added, it might make sense to target some improvements for this in a follow-up that improves the testing of the operator stuff in general.